src/index.js

```diff
@@ -25,6 +25,8 @@ export default function Knex(config) {
   let Dialect;
   if (arguments.length === 0 || (!config.client && !config.dialect)) {
     Dialect = makeClient(Client)
+  } else if (typeof config.client === 'function' && config.client.prototype instanceof Client) {
```

Comment:
@rhys-vdw I must say that you are not right. You are not passing initialized Client but the definition of the Client. Lets see it in node cli.
```bash
> Client = require('knex/lib/index').Client
{ [Function: Client] ... }

> MysqlClient = require('knex/lib/dialects/mysql')
{ [Function: Client_MySQL] ... }

> MysqlClient instanceof Client
false

> MysqlClient.prototype instanceof Client
true
```